versioned/tiered/rocksdb/src/main/java/org/projectnessie/versioned/rocksdb/RocksBaseValue.java

```diff
@@ -164,6 +175,259 @@ protected static String conditionNotMatchedMessage(Function function) {
+           */
+          @Override
+          public int compare(ExpressionPath.PathSegment ps1, ExpressionPath.PathSegment ps2) {
+            if (ps1.isName() && ps2.isName()) {
+              if (ps1.asName().getName().equals(ps2.asName().getName())) {
+                if (!ps1.getChild().isPresent() && ps2.getChild().isPresent()) {
+                  return 1;
+                } else if (ps1.getChild().isPresent() && !ps2.getChild().isPresent()) {
+                  return -1;
+                } else if (!ps1.getChild().isPresent() && !ps2.getChild().isPresent()) {
+                  return 0;
+                } else {
+                  return compare(ps1.getChild().get(), ps2.getChild().get());
+                }
+              } else {
+                return ps1.asName().getName().compareTo(ps2.asName().getName());
+              }
+            } else if (ps1.isName()) {
+              return -1;
+            } else if (ps1.isPosition() && ps2.isPosition()) {
+              return ps2.asPosition().getPosition() - ps1.asPosition().getPosition();
+            } else {
+              return 1;
+            }
+          }
+        }).collect(Collectors.toList());
+
+    for (ExpressionPath.PathSegment removedPath : removedPaths) {
+      if (removedPath.getChild().isPresent()) {
+        remove(removedPath.asName().getName(), removedPath.getChild().get());
+      }
+    }
+  }
+
+  /**
+   * Filters through the list of UpdateClause objects to remove conflicting updates. The order of
+   * the updates is preserved. Conflicts are defined as follows:
+   * <ul>
+   *   <li>Remove -&gt; Remove conflicts if latter remove is same node or a child node of former remove</li>
+   *   <li>Remove -&gt; SetEquals conflicts if SetEquals is for the same node or a child node of Remove</li>
+   *   <li>Remove -&gt; SetEquals conflicts if Remove is for a child node of SetEquals</li>
+   *   <li>Remove -&gt; Append conflicts if Append is for the same node or a child node of Remove</li>
+   *   <li>SetEquals -&gt; Remove conflicts if Remove is for a child node of SetEquals</li>
+   *   <li>SetEquals -&gt; SetEquals conflicts if latter SetEquals is for child node of the former SetEquals</li>
+   *   <li>SetEquals -&gt; Append conflicts if Append is for the same node or a child node of SetEquals</li>
+   * </ul>
+   * @param updateClauses The list of UpdateClause objects to filter
+   * @return A safe list of UpdateFunction objects to apply
+   */
+  private List<UpdateFunction> removeUnecessaryFunctions(List<UpdateClause> updateClauses) {
+    final List<UpdateFunction> updateFunctions = updateClauses
+        .stream()
+        .map(uc -> uc.accept(RocksDBUpdateClauseVisitor.ROCKS_DB_UPDATE_CLAUSE_VISITOR))
+        .collect(Collectors.toList());
+    final boolean[] keepUpdateFunction = new boolean[updateFunctions.size()];
+    Arrays.fill(keepUpdateFunction, true);
+
+    for (int i = 1; i < updateClauses.size(); i++) {
```

Comment:
Why is this starting at 1?

Reply:
I'll try to clean up these for loops and hopefully if statements. It starts at 1, since we always start by assuming that the first UpdateFunction will be applied. We then go over the remaining UpdateFunctions to see if there is a conflict with an earlier UpdateFunction.

Reply:
I didn't get a chance to try to clean this up yet.